src/RingGrids/general.jl

```diff
@@ -0,0 +1,324 @@
+
+    rings = eachring(grid)
+    for ring in rings
+        for ij in ring
+            grid[ij]"""
+@inline eachring(grid::AbstractGridArray) = grid.rings
+
+function eachring(Grid::Type{<:AbstractGridArray}, nlat_half::Integer)
+    rings = Vector{UnitRange{Int}}(undef, get_nlat(Grid, nlat_half))
+    each_index_in_ring!(rings, Grid, nlat_half)
+    return rings
+end
+
+"""
+$(TYPEDSIGNATURES)
+Same as `eachring(grid)` but performs a bounds check to assess that all grids
+in `grids` are of same size."""
+function eachring(grid1::Grid, grids::Grid...) where {Grid<:AbstractGridArray}
+    @inline 
+    n = length(grid1)
+    Base._all_match_first(X->length(X), n, grid1, grids...) || throw(BoundsError)
+    return eachring(grid1)
+end
+
+function grids_match(A::AbstractGridArray, B::AbstractGridArray)
+    length(A) == length(B) && return grids_match(typeof(A), typeof(B))
+    return false
+end
+
+function grids_match(A::Type{<:AbstractGridArray}, B::Type{<:AbstractGridArray})
+    return nonparametric_type(A) == nonparametric_type(B)
+end
+
+"""
+$(TYPEDSIGNATURES)
+UnitRange to access data on grid `grid` on ring `j`."""
+function each_index_in_ring(grid::Grid, j::Integer) where {Grid<:AbstractGridArray}
+    return each_index_in_ring(Grid, j, grid.nlat_half)
+end
+
+"""
+$(TYPEDSIGNATURES)
+UnitRange to access each grid point on grid `grid`."""
+eachgridpoint(grid::AbstractGridArray) = Base.OneTo(get_npoints(grid))
+function eachgridpoint(grid1::Grid, grids::Grid...) where {Grid<:AbstractGridArray}
+    n = length(grid1)
+    Base._all_match_first(X->length(X), n, grid1, grids...) || throw(BoundsError)
+    return eachgridpoint(grid1)
+end
+
+"""
+$(TYPEDSIGNATURES)
+Obtain ring index j from gridpoint ij and Vector{UnitRange} describing rind indices
+as obtained from eachring(::Grid)"""
+function whichring(ij::Integer, rings::Vector{UnitRange{Int}})
+    @boundscheck 0 < ij <= rings[end][end] || throw(BoundsError)
+    j = 1
+    @inbounds while ij > rings[j][end]
+        j += 1
+    end
+    return j
+end
+
+## BROADCASTING
+
+# # following https://docs.julialang.org/en/v1/manual/interfaces/#man-interfaces-broadcasting
+# import Base.Broadcast: BroadcastStyle, Broadcasted
+
+# # {1} as grids are <:AbstractVector, Grid here is the non-parameteric Grid type!
+# struct AbstractGridArrayStyle{Grid} <: Broadcast.AbstractArrayStyle{1} end
+
+# # important to remove Grid{T} parameter T (eltype/number format) here to broadcast
+# # automatically across the same grid type but with different T
+# # e.g. FullGaussianGrid{Float32} and FullGaussianGrid{Float64}
+# Base.BroadcastStyle(::Type{Grid}) where {Grid<:AbstractGridArray} =
+#     AbstractGridArrayStyle{nonparametric_type(Grid)}()
+
+# # allocation for broadcasting, create a new Grid with undef of type/number format T
+# function Base.similar(bc::Broadcasted{AbstractGridArrayStyle{Grid}}, ::Type{T}) where {Grid, T}
+#     Grid(Vector{T}(undef,length(bc)))
```

Comment:
Maybe it's already enough to change this line here to include the `ArrayType`. Maybe this needs an additional parameter in the broadcast style to distinguish between CPU and GPU style.

Reply:
I skipped the GPU broadcasting for now as I'm not sure it's actually needed? We won't be interactively allocating arrays there through broadcasting anyway maybe?

Reply:
mh..., maybe not, but we should give it a proper test if typical broadcasting operations work on GPU (or with JLArrays as well) 

So something like 
```Julia 
A .+= b 
```